File: Carros Objective-C/Carros Objective-C/Controller/MapViewController.m
 //
//  MapViewController.m
//  Carros Objective-C
//
//  Created by Antonio Lopes on 9/2/16.
//  Copyright © 2016 Antonio Lopes. All rights reserved.
//

#import "MapViewController.h"
#import "Alerta.h"

@interface MapViewController ()

@end

@implementation MapViewController

@synthesize mapView, locManager, pin;

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    locManager = [[CLLocationManager alloc] init];
    [locManager requestWhenInUseAuthorization];
    
    if (self.carro != nil) {
        double latitude = [self.carro latitude].doubleValue;
        double longitude = [self.carro longitude].doubleValue;
        
        CLLocationCoordinate2D center = CLLocationCoordinate2DMake(latitude, longitude);
        MKCoordinateSpan span = MKCoordinateSpanMake(0.01, 0.01);
        MKCoordinateRegion location = MKCoordinateRegionMake(center, span);
        
        [mapView setRegion:location animated:true];
        
        pin = [[MKPointAnnotation alloc] init];
        [pin setCoordinate:location.center];
        [pin setTitle:[NSString stringWithFormat:@"Fábrica %@", [self.carro nome]]];
        
        [mapView addAnnotation:pin];
    }
    
    self.mapView.delegate = self;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

-(nullable MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id<MKAnnotation>)annotation {
    //
    MKPinAnnotationView *pinView = [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:@"pinView"];
    [pinView setPinTintColor:[UIColor redColor]];
    [pinView setCanShowCallout:true];
    //
    UIButton *btbin = [UIButton buttonWithType:UIButtonTypeDetailDisclosure];
    [pinView setRightCalloutAccessoryView:btbin];
    //
    return  pinView;
}

- (void)mapView:(MKMapView *)mapView annotationView:(MKAnnotationView *)view calloutAccessoryControlTapped:(UIControl *)control {
    [Alerta alerta:@"Opa!" withViewController:self];
}


@end

File: Carros Swift/Carros Swift/Controller/MapViewController.swift
//
//  MapViewController.swift
//  Carros Swift
//
//  Created by Antonio Lopes on 9/1/16.
//  Copyright © 2016 Antonio Lopes. All rights reserved.
//

import UIKit
import MapKit

class MapViewController: UIViewController, MKMapViewDelegate {

    let locManager = CLLocationManager()
    var carro: Carro?
    @IBOutlet var mapView : MKMapView!
    var pin = MKPointAnnotation()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        locManager.requestWhenInUseAuthorization()
        
        if (carro != nil) {
            let lat = (self.carro!.latitude! as NSString).doubleValue
            let long = (self.carro!.longitude! as NSString).doubleValue
            
            let center = CLLocationCoordinate2D(latitude: lat, longitude: long)
            let span = MKCoordinateSpan(latitudeDelta: 0.01, longitudeDelta: 0.01)
            let location = MKCoordinateRegion(center: center, span: span)
            
            self.mapView.setRegion(location, animated: true)
            
            pin.coordinate = location.center
            pin.title = "Fábrica \(self.carro!.nome)"
            self.mapView.addAnnotation(pin)
        }
        
        mapView.delegate = self
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
        
    }
    

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

    func mapView(mapView: MKMapView, viewForAnnotation annotation: MKAnnotation) -> MKAnnotationView? {
        //
        let pinView = MKPinAnnotationView(annotation: annotation, reuseIdentifier: "pinView")
        pinView.pinTintColor = UIColor.redColor()
        pinView.canShowCallout = true
        //
        let btPin = UIButton(type: UIButtonType.DetailDisclosure) as UIView
        pinView.rightCalloutAccessoryView = btPin
        
        return pinView
    }
    
    func mapView(mapView: MKMapView, annotationView view: MKAnnotationView, calloutAccessoryControlTapped control: UIControl) {
        Alerta.alerta("Opa!", viewController: self)
    }
}
